### PR TITLE
Add intervals handling to ViewSam and associated unit test

### PR DESCRIPTION
**Pull request by abaumann**
Added ability to pass argument to ViewSam for an interval file
that the output will be restricted to.  This does not support
both the in-line intervals as well as interval file that the
GATK -L command supports.  Unit test added to verify records
from a given SAM are matched with overlapping intervals file,
and no others are output.  Updated existing TODOs to make use
of PicardCommandLine rather than direct calls into ViewSam class
within unit tests.

Resolves: #173

### CODE REVIEW

**Comment by abaumann** [comment]
Note the TODOs in ViewSam and ViewSamTest - these can be answered and modified by me before the merge

File: src/java/picard/sam/ViewSam.java [hunk 1]
```diff
@@ -75,6 +82,9 @@
     @Option(doc="Print the alignment records only.", optional = true)
     public boolean RECORDS_ONLY = false;
 
+    @Option(shortName = "int", doc = "An intervals file used to restrict what records are output.", optional = true)
+    public File INTERVAL_LIST;
```
Comment (nh13) [inline src/java/picard/sam/ViewSam.java]: I would get rid of shortname here.  Also, all options should be capitalized, even the shortname.

File: src/java/picard/sam/ViewSam.java [hunk 2]
```diff
@@ -98,11 +108,25 @@ protected int doWork() {
...
+            final CloseableIterator<SAMRecord> samRecordsIterator;
+            final SamReader samReader = SamReaderFactory.makeDefault()
+                    .referenceSequence(REFERENCE_SEQUENCE)
+                    .open(SamInputResource.of(INPUT));
+
+            // if we are only using the header or we aren't using intervals, then use the reader as the iterator.
+            if (HEADER_ONLY || INTERVAL_LIST == null) {
+                samRecordsIterator = samReader.iterator();
+            }
+            // otherwise use the SamRecordIntervalIteratorFactory to make an interval-ed iterator
```
Comment (nh13) [inline src/java/picard/sam/ViewSam.java]: move this comment to before the if as not to split the ```if/else``` block.

File: src/tests/java/picard/sam/ViewSamTest.java [hunk 3]
```diff
@@ -49,13 +61,102 @@ public void testHeaderVersion() throws Exception {
...
+        // the "filtered" file that we can compare to the intervals and ensure that only
+        // overlapped records were written
+        final SamReader samReader = SamReaderFactory.makeDefault().open(viewSamOutputFile);
+
+        // make sure the intervals file caused at least one match to be found
+        boolean foundMatches = false;
+
+        for (SAMRecord samRecord : samReader) {
+            // make an interval representing this SAM record
+            Interval samRecordInterval = new Interval(samRecord.getContig(), samRecord.getStart(), samRecord.getEnd());
```
Comment (nh13) [inline src/tests/java/picard/sam/ViewSamTest.java]: final?

File: src/tests/java/picard/sam/ViewSamTest.java [hunk 4]
```diff
@@ -49,13 +61,102 @@ public void testHeaderVersion() throws Exception {
...
+        // make a reader that is not using intervals to load the output file we wrote that
+        // was written by the call to ViewSam with the given interval file.  This will give us
+        // the "filtered" file that we can compare to the intervals and ensure that only
+        // overlapped records were written
+        final SamReader samReader = SamReaderFactory.makeDefault().open(viewSamOutputFile);
+
+        // make sure the intervals file caused at least one match to be found
+        boolean foundMatches = false;
+
+        for (SAMRecord samRecord : samReader) {
```
Comment (nh13) [inline src/tests/java/picard/sam/ViewSamTest.java]: final?

**Comment by nh13** [comment]
Assigning to @vdauwera to finalize.

**Comment by vdauwera** [comment]
What does "finalize" mean in this context, @nh13 ?

**Comment by nh13** [comment]
I suppose it means fix the TODOs?

File: src/tests/java/picard/sam/ViewSamTest.java [hunk 5]
```diff
@@ -49,13 +61,77 @@ public void testHeaderVersion() throws Exception {
...
         viewSam.INPUT = inputSam.getAbsolutePath();
+
+        // create a print stream to this file
         final PrintStream viewSamPrintStream = new PrintStream(viewSamOutputFile);
-        // TODO - Should switch over to using invocation via new PicardCommandLine() - BUT the test here is accessing class members directly.
+        // make sure the command line call exited successfully
         Assert.assertEquals(viewSam.writeSamText(viewSamPrintStream), 0);
         viewSamPrintStream.close();
+
+        viewSamPrintStream.close();
```
Comment (vdauwera) [inline src/tests/java/picard/sam/ViewSamTest.java]: Isn't this line a duplicate of L 72?

File: src/java/picard/sam/ViewSam.java [hunk 6]
```diff
@@ -98,11 +108,24 @@ protected int doWork() {
...
+
+            // if we are only using the header or we aren't using intervals, then use the reader as the iterator.
+            // otherwise use the SamRecordIntervalIteratorFactory to make an interval-ed iterator
+            if (HEADER_ONLY || INTERVAL_LIST == null) {
+                samRecordsIterator = samReader.iterator();
+            } else {
+                IOUtil.assertFileIsReadable(INTERVAL_LIST);
+
+                final List<Interval> intervals = IntervalList.fromFile(INTERVAL_LIST).uniqued().getIntervals();
+                // TODO: what to do about the useIndex parameter?
```
Comment (vdauwera) [inline src/java/picard/sam/ViewSam.java]: I can't answer this.

Comment (nh13) [inline src/java/picard/sam/ViewSam.java]: Can you get the answer from @abaumann?

**Comment by vdauwera** [comment]
Someone else needs to answer that last TODO, @nh13. I have no clue.

**Comment by abaumann** [comment]
The TODOs I put in because I couldn't answer them - I can look tomorrow to
see if I can answer them not in hindsight and if not then someone else
familiar with Picard may be able to answer them.  I was just concerned
about certain implications of code I didn't fully understand and added
TODOs for this uncertainty.  That line is a duplicate, I'll fix that.

File: src/java/picard/sam/ViewSam.java (hunk 6 above)
Comment (nh13) [inline src/java/picard/sam/ViewSam.java]: What about using ```samReader.hasIndex()```?

**Comment by nh13** [comment]
@abaumann I think there was only one TODO left, and hopefully I gave you the answer.

**Comment by abaumann** [comment]
should be good now, i followed the suggested samReader.hasIndex() which appears to work

**Pull request merged**